sdks/java/extensions/histogram/src/main/java/org/apache/beam/sdk/extensions/histogram/Histogram.java

```diff
@@ -0,0 +1,571 @@
+    private static final VarIntCoder INT_CODER = VarIntCoder.of();
+
+    /**
+     * Index to indicate method used where only non-empty buckets are encoded with their indices.
+     */
+    private static final int NON_EMPTY_BUCKETS_CODER = 0;
+
+    /** Index to indicate method used where all buckets are encoded sequentially. */
+    private static final int ALL_BUCKETS_CODER = 1;
+
+    /** Encoded size of 0 in bytes. */
+    private static final long ENCODED_ZERO_SIZE = VarInt.getLength(0);
+
+    @Override
+    public void encode(HistogramAccumulator value, OutputStream outStream) throws IOException {
+      checkNotNull(value, "can not encode a null histogram.");
+
+      int numEmptyBucketsAtTheEnd = 0;
+      for (int i = value.counts.length - 1; i >= 0; i--) {
+        if (value.counts[i] != 0) {
+          break;
+        }
+        numEmptyBucketsAtTheEnd++;
+      }
+
+      if (shouldEncodeNonEmptyBucketsOnly(value, numEmptyBucketsAtTheEnd)) {
+        encodeNonEmptyBuckets(value, numEmptyBucketsAtTheEnd, outStream);
+      } else {
+        encodeAllBuckets(value, numEmptyBucketsAtTheEnd, outStream);
+      }
+    }
+
+    /**
+     * Estimates the size of the accumulator and returns whether encoding only non-empty buckets
+     * with indices produces a smaller accumulator.
+     *
+     * <p>To check the difference in sizes between these two encoding methods, we do not add the
+     * size of non-empty encoded bucket count because it is added to both, and we just cancel it for
+     * both sides. So, comparison is done between sizes of encoded indices and sizes of intermediate
+     * zero counts (not trailing zeros).
+     */
+    private boolean shouldEncodeNonEmptyBucketsOnly(
+        HistogramAccumulator value, int numEmptyBucketsAtTheEnd) {
+      int numBuckets = value.counts.length;
+      long nonEmptyBucketIndicesEncodedSize = 0;
+      long numZeroBuckets = 0;
+
+      for (int j = 0; j < numBuckets - numEmptyBucketsAtTheEnd; j++) {
+        if (value.counts[j] != 0) {
+          nonEmptyBucketIndicesEncodedSize += VarInt.getLength(j);
+        } else {
+          numZeroBuckets++;
+        }
+      }
+
+      return nonEmptyBucketIndicesEncodedSize < numZeroBuckets * ENCODED_ZERO_SIZE;
+    }
+
+    private void encodeNonEmptyBuckets(
+        HistogramAccumulator value, int numEmptyBucketsAtTheEnd, OutputStream outStream)
+        throws IOException {
+      // As we have two different encoding methods, the first byte indicates the coder used.
+      outStream.write(NON_EMPTY_BUCKETS_CODER);
+      int numBuckets = value.counts.length;
+      INT_CODER.encode(numBuckets, outStream);
+
+      List<Integer> indices = new ArrayList<>();
+      List<Long> counts = new ArrayList<>();
+
+      for (int i = 0; i < numBuckets - numEmptyBucketsAtTheEnd; i++) {
+        if (value.counts[i] != 0) {
+          indices.add(i);
+          counts.add(value.counts[i]);
+        }
+      }
```

Comment:
I think we don't need to store the index of non-zero bucket in an intermediate list, we could just directly emit the (index, value) pair into the outStream, since we already know the number of elements that need to be encoded because it was calculated earlier when choosing which coding method to use, but again - maybe it's not worth to have this optimization

Reply:
the count of non-empty buckets is not directly done in the encode method, but inside the shouldEncodeNonEmptyBucketsOnly() additional method which returns the boolean for whether to encode non-empty buckets only. So, we thought it is better to first insert into a separate list, and only after encode index and count in the separate for-loop with less runtime(at least <=N where N is numBuckets without trailing zeros).